### PR TITLE
Feature/tool search page

### PR DESCRIPTION
**Pull request by Ankur1493**
**How the UI looks**
-- why this UI --> The ui is inspired from the landing page of devToolsAcademy, It can be done a lot better...

image
image

how it looks on phone
image

**Demo**
[Screencast from 13-11-24 01:39:57 AM IST.webm](https://github.com/user-attachments/assets/7d849a02-3b32-4f47-afea-aedd4fc7442d)

**Features added** - 
 - Search with debouncing
 - Loading and Errors managed
 - Filters based on licenses, categories etc
 - Can be sorted on last updated and github stars
 - Responsive UI

**What could have been done better** - 
I should have integrated a zustand state to organize the tools better and used that for filtering and sorting the data.
If zustand would have been added then the compare and other parts would have become easier, with adding infinte scroll and all as search could have been just for searching among all.

**Why no bonus features added as of now???**
- I am a bit busy with the launch of my own saas product in this month, and I wanted to see is there any changes you want, like if you wanted or allow me to introduce zustand I would have done a better job

**How to run this** 
- It's simple just do an npm install, and I had to add clerk details so you might have to add them, else rest is running pretty normal

**Future Improvements**
 - Add a zustand state
 - Add infinite scroll and create sorting and filter seperately using state
 - make a seperate tool detail page, and showcase only some details of tool on cards
 - Add bonus features

## Release Notes

- **New Features**
	- Introduced a new tool filtering component, allowing users to filter tools by categories, tags, and licenses.
	- Added a new API endpoint to retrieve and filter tools based on user-defined criteria.
	- Implemented a user-friendly "No tools found" message for better user experience.
	- Added a loading skeleton to indicate loading states for tool data.
	- New tools display components (ToolCard, ToolsGrid) for enhanced presentation of tool information.

- **Improvements**
	- Enhanced button and input components with updated styles for better usability.
	- Introduced a custom Select component for improved dropdown functionality.

- **Bug Fixes**
	- Resolved issues with form handling and validation through the addition of new form components.

### CODE REVIEW

File: src/components/ui/select.tsx [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+      ref={ref}
+      className={cn(
+        "relative z-50 max-h-96 min-w-[8rem] overflow-hidden rounded-md border bg-popover text-popover-foreground shadow-md data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2",
+        position === "popper" &&
+          "data-[side=bottom]:translate-y-1 data-[side=left]:-translate-x-1 data-[side=right]:translate-x-1 data-[side=top]:-translate-y-1",
+        className
+      )}
+      position={position}
+      {...props}
+    >
```
[the remark this answers is not in the document]
Comment (Ankur1493) [reply src/components/ui/select.tsx:85]: shadcn component

File: src/data/tools.ts [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+    "websiteUrl": "https://www.sonarqube.org/",
+    "tags": [
+      "code quality",
+      "security",
+      "static analysis"
+    ],
+    "lastUpdate": "2023-10-25",
+    "license": "LGPLv3"
+  }
+]
```
[the remark this answers is not in the document]
Comment (Ankur1493) [reply src/data/tools.ts:383]: normal data file, just written here for the demo

File: src/components/tools/ToolsGrid.tsx [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        search: searchTerm,
+        sortBy,
+        tags: searchParams.getAll('tags'),
+        categories: searchParams.getAll('categories'),
+        licenses: searchParams.getAll('licenses')
+      }
+
+      const response = await axios.get(`${process.env.NEXT_PUBLIC_BASE_URL}/api/tools`, {
+        params
+      })
```
[the remark this answers is not in the document]
Comment (Ankur1493) [reply src/components/tools/ToolsGrid.tsx:43]: I have defined it in .env, it's simply localhost:3000

**Comment by tyaga001** [comment]
@Ankur1493, as discussed, please share a detailed proposal of how you would like to proceed with all the features development to build the complete e2e directory.

**Pull request closed without merging**